File: C#/Queue/Queue.cs
namespace Queue;

class Queue<T>
{
    private class Node
    {
        public T data { get; set; }

        public Node next { get; set; }

        public Node(T data, Node next)
        {
            this.data = data;
            this.next = next;
        }
    }

    private Node head;
    private Node tail;

    private int count;

    public Queue()
    {
        head = tail = null;
        count = 0;
    }

    public void Enqueue(T item)
    {
        if(tail == null)
        {
            head = tail = new Node(item, null);
        }
        else
        {
            tail.next = new Node(item, null);
            tail = tail.next;
        }
        count ++;
    }

    public T Dequeue()
    {
        if(head == null)
        {
            return default(T);
        }

        T result = head.data;
        head = head.next;
        return result;
    }

    public T Peek()
    {
        if(head == null)
        {
            return default(T);
        }
        return head.data;
    }

}

File: JS/queue.js
class Node{
  constuctor(value){
    this.value = value;
    this.next = null;
  }
}

class Queue {
  constuctor(){
    this.first = null;
    this.last = null;
    this.size = 0;
  }
  enqueue(val){
    var newNode = new Node(val);
    if(!this.first){
      this.first = newNode;
      this.last = newNode;
    }
    else{
      this.last.next = newNode;
      this.last = newNode;
    }
    return ++this.size;

  }
  dequeue(){
    if(!this.first) return null;

    var temp = this.first;
    if(this.first === this.last){
      this.last = null;
    }
    this.first = this.first.next;
    this.size--;
    return temp.value;

  }
}

var q = new Queue();
q.enqueue("first");
q.enqueue("second");
console.log(q);
q.dequeue();
console.log(q);
